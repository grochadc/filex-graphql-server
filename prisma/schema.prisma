generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("HEROKU_POSTGRESQL_TEAL_URL")
}

model Student {
  id                  Int                   @id(map: "student_pkey1") @default(autoincrement())
  applicantId         Int                   @map("applicant_id")
  applicant           Applicant             @relation(fields: [applicantId], references: [id])
  ciclo_actual        String                @db.VarChar(255)
  codigo              String                @db.VarChar(255)
  nivel               Int
  curso               String                @db.VarChar(255)
  situacion           String?               @db.VarChar(255)
  groupId             Int?                  @map("group_id")
  groupObject         Group?                @relation(fields: [groupId], references: [id])
  desertor            Boolean?
  WorkshopReservation WorkshopReservation[]

  @@map("student")
}

model TestResults {
  id                 Int      @id @default(autoincrement())
  codigo             String   @db.VarChar(255)
  nombre             String   @db.VarChar(255)
  apellidoPaterno    String   @map("apellido_paterno") @db.VarChar(255)
  apellidoMaterno    String   @map("apellido_materno") @db.VarChar(255)
  genero             String   @db.VarChar(255)
  ciclo              String   @db.VarChar(255)
  carrera            String   @db.VarChar(255)
  telefono           String   @db.VarChar(255)
  email              String   @db.VarChar(255)
  institucionalEmail String?  @map("institucionalemail") @db.VarChar(255)
  nivelEscrito       Int      @map("nivel_escrito")
  curso              String   @db.VarChar(255)
  externo            Boolean?
  reubicacion        Boolean?
  generated_id       String?  @db.VarChar(255)
  meetLink           String?  @map("meetlink") @db.VarChar(255)
  nivelOral          Int?     @map("nivel_oral")
  nivelFinal         Int?     @map("nivel_final")

  @@map("test_results")
}

model Teacher {
  id      Int              @id @default(autoincrement())
  nombre  String           @db.VarChar(255)
  groups  Group[]
  options WorkshopOption[]

  @@map("teacher")
}

model Applicant {
  id                  Int       @id @default(autoincrement())
  codigo              String?   @db.VarChar(255)
  nombre              String?   @db.VarChar(255)
  apellido_paterno    String?   @db.VarChar(255)
  apellido_materno    String?   @db.VarChar(255)
  genero              String?   @db.VarChar(255)
  carrera             String?   @db.VarChar(255)
  ciclo_ingreso       String?   @db.VarChar(255)
  telefono            String?   @db.VarChar(255)
  email               String?   @db.VarChar(255)
  institutional_email String?   @db.VarChar(255)
  externo             Boolean?
  student             Student[]

  @@map("applicant")
}

model Group {
  ciclo       String    @db.VarChar(255)
  name        String    @db.VarChar(255)
  time        String    @db.VarChar(255)
  aula        String    @db.VarChar(255)
  teacherId   Int       @map("teacher_id")
  teacher     Teacher   @relation(fields: [teacherId], references: [id])
  id          Int       @id @default(autoincrement())
  nivel       Int
  registrados Int       @default(0)
  course      String?   @db.VarChar(255)
  student     Student[]

  @@map("group_table")
}

model student_ciclo {
  id               Int     @id(map: "student_pkey") @default(autoincrement())
  codigo           String
  nombre           String
  apellido_paterno String
  apellido_materno String
  genero           String
  carrera          String
  ciclo_ingreso    String?
  telefono         String
  email            String
  nivel            String
  curso            String
  externo          Boolean
  grupo            String
}

model student_old {
  id           Int     @id(map: "course_pkey") @default(autoincrement())
  codigo       String  @db.VarChar(255)
  nivel        String? @db.VarChar(255)
  curso        String? @db.VarChar(255)
  grupo        String? @db.VarChar(255)
  ciclo_actual String? @db.VarChar(255)
  applicant_id Int?
  situacion    String? @db.VarChar(255)
}

model WorkshopOption {
  id                  Int                   @id @default(autoincrement())
  active              Boolean
  day                 String?               @db.VarChar(255)
  time                String?               @db.VarChar(255)
  teacher_id          Int
  teacher             Teacher               @relation(fields: [teacher_id], references: [id])
  workshop_id         Int
  workshop            Workshop              @relation(fields: [workshop_id], references: [id])
  url                 String?               @db.VarChar(255)
  place               String?               @db.VarChar(255)
  workshopReservation WorkshopReservation[]

  @@map("workshop_option_info")
}

model WorkshopReservation {
  id          Int            @id @default(autoincrement())
  create_time DateTime       @default(dbgenerated("CURRENT_DATE")) @db.Date
  student_id  Int
  student     Student        @relation(fields: [student_id], references: [id])
  option_id   Int
  option      WorkshopOption @relation(fields: [option_id], references: [id])
  attended    Boolean?

  @@map("workshop_reservation")
}

model Workshop {
  id          Int              @id @default(autoincrement())
  name        String           @db.VarChar(255)
  description String           @db.VarChar(255)
  levels      Int[]
  options     WorkshopOption[]

  @@map("workshop")
}
