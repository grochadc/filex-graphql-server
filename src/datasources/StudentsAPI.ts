import { RESTDataSource } from "apollo-datasource-rest";

export type GetStudentFn = (codigo: string, env?: 'dev'|'prod') => Promise<any>;

class StudentsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://filex-5726c.firebaseio.com/students";
  }

  async _get(url: string, env:string|undefined) {
    const finalUrl = env ? '/'+env+url : '/prod'+url;
    return this.get(finalUrl);
  }

  async getStudent(...[codigo, env]: Parameters<GetStudentFn>) {
    return this.get(`/${env}/${codigo}.json`);
  }
}

export { StudentsAPI };
