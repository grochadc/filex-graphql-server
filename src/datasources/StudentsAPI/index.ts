import { DataSource } from "apollo-datasource";
import { StudentModel } from "../../modules/workshops/models";
import { ParameterizedQuery as PQ } from "pg-promise";
import { ApolloError } from "apollo-server";
import { PrismaClient } from "@prisma/client";
import {
  StudentInput,
  ApplicantInput,
  Applicant,
} from "../../generated/graphql";

type MyStudentInput = Omit<StudentModel, "id">;
type Changes = Partial<MyStudentInput>;

class StudentsAPI extends DataSource {
  db: any;
  prisma: PrismaClient;
  constructor(db: any, prisma: PrismaClient) {
    super();
    if (prisma === undefined)
      throw new Error("Include a new PrismaCLient() on constructor");
    this.prisma = prisma;
    this.db = db;
    if (prisma === undefined) {
      throw Error("Supply a new PrismaCLient() on constructor");
    }
    this.prisma = prisma;
  }

  async getAllStudents(cicloActual: string) {
    const students = await this.prisma.student.findMany({
      where: {
        cicloActual: cicloActual,
        groupId: {
          not: null
        }
      },
      include: {
        applicant: true,
        groupObject: {
          include: {
            teacher: true,
          },
        },
      },
    });

    return students;
  }

  async getStudent(codigo: string, cicloActual: string) {
    if (codigo === null)
      throw new Error("Codigo not provided on studentsAPI.getStudent");
    if (cicloActual === null)
      throw new Error("ciclo_actual not provided on studentsAPI.getStudent");
    const student = await this.prisma.student.findFirst({
      include: {
        applicant: true,
        groupObject: true,
      },
      where: {
        codigo,
        cicloActual,
      },
    });
    if (student === null) throw new Error("Student not found on database");
    if (student.applicant == null)
      throw new Error("Applicant info is missing on database");
    if (student.groupObject == null)
      throw new Error("Couldnt find a group Object");

    return student;
  }

  async addApplicant(
    applicant: ApplicantInput
  ): Promise<Omit<Applicant, "curso" | "groups" | "nivel" | "registering">> {
    const result = await this.prisma.applicant.create({
      data: applicant,
    });

    return {
      ...result,
      id: String(result.id),
    };
  }

  async addStudent(student: StudentInput) {
    const applicant = await this.prisma.applicant.findFirst({
      where: {
        codigo: student.codigo,
      },
    });

    if (applicant === null)
      throw new Error(
        "No applicant found on database. Please submit the student's personal information on the Applicant endpoint first."
      );

    const groupObj = await this.prisma.group.findFirst({
      where: {
        name: student.grupo,
        ciclo: student.cicloActual,
      },
    });
    return Promise.resolve();
    /*
    return this.prisma.student.create({
      data: {
        applicant: { connect: {id: applicant.id} },
        codigo: student.codigo,
        curso: student.curso,
        groupObject: { connect: { id: groupObj.id } },
        ciclo_actual: student.cicloActual,
        nivel: student.nivel
      }
    })
    */
  }

  async editStudent(codigo: string, changes: Changes, ciclo: string) {
    const oldStudent = await this.getStudent(codigo, ciclo);
    const newStudent = Object.assign(oldStudent, changes);
    try {
      this.db.none(
        new PQ({
          text: EDIT_STUDENT,
          values: [
            newStudent.codigo,
            newStudent.nombre,
            newStudent.apellido_materno,
            newStudent.apellido_paterno,
            newStudent.genero,
            newStudent.carrera,
            newStudent.ciclo,
            newStudent.telefono,
            newStudent.email,
            newStudent.nivel,
            newStudent.grupo,
            newStudent.externo,
            newStudent.curso,
            newStudent.id,
          ],
        })
      );
    } catch (e) {
      console.error(e);
    }
    return newStudent;
  }
}

export const GET_STUDENT = `
  SELECT * FROM Student WHERE codigo=$1
`;

export const ADD_STUDENT = `
  INSERT INTO Student (
    codigo,
    nombre,
    apellido_materno,
    apellido_paterno,
    genero,
    carrera,
    ciclo,
    telefono,
    email,
    nivel,
    grupo,
    externo,
    curso
  ) values (
    $1,
    $2,
    $3,
    $4,
    $5,
    $6,
    $7,
    $8,
    $9,
    $10,
    $11,
    $12,
    $13
  )
  RETURNING id
`;

export const EDIT_STUDENT = `
  UPDATE Student SET
  codigo=$1,
  nombre= $2,
  apellido_materno= $3,
  apellido_paterno= $4,
  genero= $5,
  carrera= $6,
  ciclo= $7,
  telefono= $8,
  email= $9,
  nivel= $10,
  grupo= $11,
  externo= $12,
  curso= $13
  WHERE id=$14
`;

export { StudentsAPI };
