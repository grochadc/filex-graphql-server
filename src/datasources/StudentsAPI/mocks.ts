import { ParameterizedQuery as PQ } from "pg-promise";
import { ADD_STUDENT, GET_STUDENT } from "./index";

export const applicants = [
  {
    id: 1,
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    carrera: "Abogado",
    ciclo_ingreso: "2021A",
    telefono: "1234567890",
    email: "bad@bunny.pr",
    externo: false,
  },
  {
    id: 2,
    codigo: "0987654321",
    nombre: "Alberto",
    apellido_paterno: "Aguilera",
    apellido_materno: "Valadez",
    genero: "M",
    carrera: "Abogado",
    ciclo_ingreso: "2021A",
    telefono: "1234567890",
    email: "juanga@elnoanoa.mx",
    externo: false,
  },
];

export const students = [
  {
    id: 1,
    applicantId: 1,
    ciclo_actual: "2022B",
    codigo: "1234567890",
    nivel: 4, 
    curso: "en",
    groupId: 1,
    desertor: false,
    situacion: null,
  },
  {
    id: 2,
    applicantId: 2,
    ciclo_actual: "2022B",
    codigo: "0987654321",
    nivel: 3, 
    curso: "en",
    groupId: 2,
    desertor: false,
    situacion: null,
  }
]

const groups = [
  {
    id: 1,
    ciclo: "2022B",
    name: "E4-1",
    time: "07:00-08:00",
    aula: "G5",
    teacherId: 1,
    nivel: 4,
    registrados: 1,
    course: "en"
  },
  {
    id: 2,
    ciclo: "2022B",
    name: "E3-1",
    time: "07:00-08:00",
    aula: "G2",
    teacherId: 2,
    nivel: 3,
    registrados: 1,
    course: "en"
  }
];


export const studentss = [
  {
    id: 1,
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    carrera: "Abogado",
    ciclo: "2021A",
    telefono: "1234567890",
    email: "bad@bunny.pr",
    nivel: "4",
    curso: "en",
    grupo: "E4-1",
    externo: false,
  },
  {
    id: 2,
    codigo: "0987654321",
    nombre: "Alberto",
    apellido_paterno: "Aguilera",
    apellido_materno: "Valadez",
    nivel: "1",
    grupo: "E1-1",
    genero: "M",
    carrera: "Abogado",
    ciclo: "2021A",
    telefono: "1234567890",
    email: "juanga@elnoanoa.mx",
    curso: "en",
    externo: false,
  },
];

const mocks = {
  db: {
    one: jest.fn((pq: PQ) => {
      if (pq.text === ADD_STUDENT) return Promise.resolve({ id: 1 });
    }),
    none: jest.fn((pq) => {
      return Promise.resolve();
    }),
    oneOrNone: jest.fn((pq) => {
      if (pq.text === GET_STUDENT) {
        const student = students.find(
          (student) => student.codigo === pq.values[0]
        );
        return Promise.resolve(student ? student : null);
      }
    }),
    any: jest.fn((pq) => {
      if (pq.text === GET_STUDENT) {
        const student = students.find(
          (student) => student.codigo === pq.values[0]
        );
        return Promise.resolve(student ? student : null);
      }
    }),
  },
};

export default mocks;
