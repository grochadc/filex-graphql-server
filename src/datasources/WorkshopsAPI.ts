import { RESTDataSource } from "apollo-datasource-rest";
import * as R from "ramda";
import * as utils from "../utils";
import database, { Option } from "./db";
import { Workshop, AttendingStudent } from "../generated/graphql";
import {
  WorkshopModel,
  TeacherModel,
  OptionModel,
  StudentReservationModel,
  TeacherReservationModel,
  DatabaseModel,
} from "../modules/workshops/models";

type Maybe<T> = T | null;

class WorkshopsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://filex-5726c.firebaseio.com/workshops/";
  }

  async isOpen(): Promise<boolean> {
    const open = await this.get(
      `${this.context.enviroment}/system/isOpen/result.json`
    );
    return open ? open : false;
  }

  async toggleOpen(): Promise<boolean> {
    const open = await this.isOpen();
    await this.put(`${this.context.enviroment}/system/isOpen.json`, {
      result: !open,
    });
    return !open;
  }

  async getWorkshops(): Promise<WorkshopModel[]> {
    const workshops = await this.get(
      `${this.context.enviroment}/workshops.json`
    );
    if (workshops === null)
      throw new Error(
        `${this.context.enviroment}/workshops.json returned null`
      );
    return Object.values(workshops);
  }

  async getOptions(): Promise<DatabaseModel["options"]> {
    return (await this.get(
      `${this.context.enviroment}/options.json`
    )) as DatabaseModel["options"];
  }

  getAvailableOptions(): Promise<Maybe<DatabaseModel["availableOptions"]>> {
    return this.get(
      `${this.context.enviroment}/availableOptions.json`
    ) as Promise<Maybe<DatabaseModel["availableOptions"]>>;
  }

  getTeacher(id: string): Promise<TeacherModel> {
    return this.get(`${this.context.enviroment}/teachers/${id}.json`);
  }

  getStudentReservation(codigo: string): Promise<StudentReservationModel> {
    return this.get(
      `/${this.context.enviroment}/studentsReservations/${codigo}.json`
    );
  }

  async makeReservation(
    codigo: string,
    teacher_id: string,
    option_id: string,
    tutorial_reason?: string
  ) {
    const option = await this.getOptionById(option_id);
    if (option === null)
      throw new Error("option with id " + option_id + " returned null");
    const student = await this.context.dataSources.studentsAPI.getStudent(
      codigo
    );
    const { nombre, apellido_paterno, apellido_materno, nivel, grupo } =
      student;
    const teacherReservation: TeacherReservationModel = {
      codigo: student.codigo,
      nombre,
      apellido_paterno,
      apellido_materno,
      nivel,
      grupo,
      option_id: option.id,
      option_name: "why? option_name from makeReservation dataSource",
      workshop_id: option.workshop_id,
      workshop_name: option.workshop_name,
      tutorial_reason: tutorial_reason ? tutorial_reason : null,
    };
    this.post(
      `${this.context.enviroment}/teachers/${teacher_id}/raw_reservations/${option_id}.json`,
      teacherReservation
    );
    this.put(`${this.context.enviroment}/studentsReservations/${codigo}.json`, {
      option_id: option_id,
    });

    //enqueue
    const availableNumber = await this.get(
      `${this.context.enviroment}/availableOptions/${option_id}.json`
    );
    this.put(
      `${this.context.enviroment}/availableOptions/${option_id}.json`,
      availableNumber + 1
    );

    return { ...option, option_id: option.id };
  }

  deleteReservations(teacher_id: string, option_id: string) {
    return this.delete(
      `${this.context.enviroment}/teachers/${teacher_id}/raw_reservations/${option_id}.json`
    );
  }

  async resetReservations() {
    await this.delete(`${this.context.enviroment}/studentsReservations.json`);
    await this.delete(`${this.context.enviroment}/availableOptions.json`);
    return true;
  }

  getOptionById(id: string): Promise<OptionModel> {
    return this.get(`${this.context.enviroment}/options/${id}.json`);
  }

  async saveAttendance(attendance: AttendingStudent[]) {
    const date = new Date();
    const values = attendance.map((student) => {
      return [
        `=date(${date.getFullYear()},${date.getMonth() + 1},${date.getDate()})`,
        student.codigo,
        student.nombre,
        student.apellido_paterno,
        student.apellido_materno,
        student.nivel,
        student.grupo,
        student.workshop,
        student.teacher,
        student.attended,
      ];
    });
    const range = "Attendance!A1";

    try {
      await this.context.dataSources.workshopsSheetsAPI.append(values, range);
    } catch (e) {
      console.error(e);
    }
    return true;
  }

  setWorkshopLink(option_id: string, url: string) {
    this.put(`${this.context.enviroment}/options/${option_id}/url.json`, url);
    return true;
  }

  async getMaxStudentReservations() {
    const result = await this.get(`${this.context.enviroment}/system.json`);
    return result.max_reservations;
  }

  async getReservationLimit(): Promise<number> {
    return this.get("/reservationLimit.json");
  }
}

export { WorkshopsAPI };
