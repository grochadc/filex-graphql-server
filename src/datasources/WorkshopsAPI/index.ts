import { RESTDataSource } from "apollo-datasource-rest";
import { AttendingStudent } from "../../generated/graphql";
import {
  WorkshopModel,
  OptionModel,
  DatabaseModel,
} from "../../modules/workshops/models";
import { PrismaClient } from "@prisma/client";
import { ApolloError } from "apollo-server";
import { students } from "datasources/StudentsAPI/mocks";

type Maybe<T> = T | null;

class WorkshopsAPI extends RESTDataSource {
  prisma: PrismaClient;
  constructor(prisma: PrismaClient) {
    super();
    if (prisma === undefined)
      throw new Error("Must include a new PrismaClient() on constrctor");
    this.prisma = prisma;
    this.baseURL = "https://filex-5726c.firebaseio.com/workshops/";
    if (process.env.NODE_ENV === "test") this.initialize({} as any);
  }

  getTodaysDate(): Date {
    return new Date();
  }

  async isOpen(): Promise<boolean> {
    const open = await this.get(
      `${this.context.enviroment}/system/isOpen/result.json`
    );
    return open ? open : false;
  }

  async toggleOpen(): Promise<boolean> {
    const open = await this.isOpen();
    await this.put(`${this.context.enviroment}/system/isOpen.json`, {
      result: !open,
    });
    return !open;
  }

  getTeacher(id: string) {
    return this.prisma.teacher.findUnique({
      where: {
        id: Number(id),
      },
    });
  }

  async getAllTeachers() {
    return this.prisma.teacher.findMany({
      include: {
        options: {
          include: {
            workshop: true,
          },
        },
      },
    });
  }

  async getReservationsByOptionId(option_id: string) {
    const openDate = await this.getOpenDate();
    const reservations = await this.prisma.workshopReservation.findMany({
      where: {
        option_id: Number(option_id),
        create_time: {
          gte: new Date(openDate),
        },
      },
      include: {
        student: {
          include: {
            applicant: true,
          },
        },
        option: true,
      },
    });

    if (null) return [];

    return reservations.map((reservation) => {
      return {
        ...reservation,
        id: String(reservation.id),
      };
    });
  }

  async getReservationCount(student_id: string) {
    return this.prisma.workshopReservation.count({
      where: {
        student_id: Number(student_id),
      },
    });
  }

  async getOptionReservationCount(option_id: string) {
    const openDate = await this.getOpenDate();

    return this.prisma.workshopReservation.count({
      where: {
        option_id: Number(option_id),
        create_time: {
          gte: new Date(openDate),
        },
      },
    });
  }

  async getAllOptions(max_students: number) {
    return this.prisma.workshopOption.findMany();
  }

  async getAllWorkshops(max_students: number) {
    return this.prisma.workshop.findMany({
      include: {
        options: {
          include: {
            teacher: true,
          },
        },
      },
    });
  }

  async getTeacherReservations(teacherId: number, optionId: number) {
    const openDateStr = await this.getOpenDate();
    return this.prisma.workshopReservation.findMany({
      where: {
        option: {
          teacher_id: teacherId,
          id: optionId,
        },
        create_time: {
          gt: new Date(openDateStr),
        },
      },
      include: {
        option: true,
        student: {
          include: {
            applicant: true,
            groupObject: true,
          },
        },
      },
    });
  }

  async getTeacherOptions(teacherId: string) {
    return this.prisma.workshopOption.findMany({
      include: {
        workshop: true,
      },
      where: {
        teacher_id: Number(teacherId),
      },
    });
  }

  async getStudentReservation(studentId: string) {
    const openDateString = await this.getOpenDate();
    const reservation = await this.prisma.workshopReservation.findFirst({
      where: {
        student_id: Number(studentId),
        create_time: {
          gte: new Date(openDateString),
        },
      },
      include: {
        student: true,
        option: {
          include: {
            workshop: true,
            teacher: true,
          },
        },
      },
    });
    if (!reservation) return null;
    return { ...reservation, id: String(reservation.id) };
  }

  async makeReservation(student_id: string, option_id: string) {
    if (!student_id)
      throw new Error(
        "No student_id provided to make a reservation on workshopsAPI.makeReservation"
      );
    if (!option_id)
      throw new Error(
        "No option_id provided to make a reservation on workshopsAPI.makeReservation"
      );
    const currentReservation = await this.getStudentReservation(student_id);
    if (currentReservation)
      throw new ApolloError("Student already has a reservation", "409");
    const reservation = await this.prisma.workshopReservation.create({
      data: {
        student_id: Number(student_id),
        option_id: Number(option_id),
      },
      include: {
        student: true,
        option: {
          include: {
            teacher: true,
            workshop: true,
          },
        },
      },
    });
    return { ...reservation, id: String(reservation.id) };
  }

  async resetReservations() {
    const now = await this.getTodaysDate();
    const nowString = now.toISOString();
    await this.put("system/openDate.json", { result: nowString });
    return true;
  }

  async getOpenDate(): Promise<string> {
    const openDate = await this.get("system/openDate.json");
    const dateString = openDate.result;
    return dateString;
  }

  async saveAttendance(attendance: AttendingStudent[]) {
    const reservation_ids = attendance
      .filter((student) => student.attended)
      .map((student) => Number(student.reservation_id));
    const modified = await this.prisma.workshopReservation.updateMany({
      where: {
        id: {
          in: reservation_ids,
        },
      },
      data: {
        attended: true,
      },
    });
    return modified.count > 0;
  }

  async getMaxStudentReservations() {
    const result = await this.get(`system.json`);
    return result.max_reservations;
  }

  async getReservationLimit(): Promise<number> {
    return this.get("/reservationLimit.json");
  }
}

export { WorkshopsAPI };
