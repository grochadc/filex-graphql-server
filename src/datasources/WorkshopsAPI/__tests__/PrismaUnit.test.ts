import { WorkshopsAPI } from "../../WorkshopsAPI";
import { StudentsAPI } from "../../StudentsAPI";
import { prismaMock } from "../../testutils/prisma/singleton";
import {
  WorkshopReservation,
  Student,
  Workshop,
  WorkshopOption,
  Teacher,
} from "@prisma/client";

type MakeReservationResult = WorkshopReservation & {
  student: Student;
  option: WorkshopOption & {
    teacher: Teacher;
    workshop: Workshop;
  };
};

const dataSources: { workshopsAPI: WorkshopsAPI; studentsAPI: StudentsAPI } =
  {} as any;

beforeAll(() => {
  dataSources.workshopsAPI = new WorkshopsAPI(prismaMock);
  dataSources.studentsAPI = new StudentsAPI({} as any, prismaMock);
});

const option_1: WorkshopOption = {
        id: 1,
        active: true,
        day: "Lunes",
        time: "13:00",
        teacher_id: 1,
        workshop_id: 1,
        url: null,
        place: "Centro de Aprendizaje Global",
      };

      const teacher_1: Teacher = {
          id: 1,
          nombre: "Gonzalo Rocha",
      };

      const workshop_1: Workshop = {
          id: 1,
          name: "Conversation",
          description: "Conversation Club",
          levels: [1, 2, 3, 4, 5, 6],
        };

describe("api.makeReservation", () => {
  test("makes a reservation", async () => {
    const student_id = 1;
    const option_id = 1;
    
    type MakeReservationResult = WorkshopReservation & {
      student: Student;
      option: WorkshopOption & {
        teacher: Teacher;
        workshop: Workshop;
      };
    };

    const mockResVal: MakeReservationResult = {
      id: 1,
      create_time: new Date(),
      student_id: Number(student_id),
      student: {
        id: 1,
        applicantId: 1,
        ciclo_actual: "2022A",
        codigo: "1234567890",
        nivel: 3,
        situacion: "AC",
        curso: "en",
        groupId: 1,
        desertor: false,
      },
      option_id: Number(option_id),
      option: {
        ...option_1,
        teacher: teacher_1,
        workshop: workshop_1,
      },
      attended: false,
    };

    //@ts-ignore
    prismaMock.workshopReservation.create.mockResolvedValue(mockResVal);

    let reservation = await dataSources.workshopsAPI.makeReservation(
      student_id,
      option_id
    );

    expect(reservation.option.id).toBe(Number(option_id));
    expect(reservation.student.id).toBe(Number(student_id));
    expect(reservation).toMatchSnapshot({
      create_time: expect.any(Date),
    });
  });

  it("fails on a null student_id", async () => {
    await expect(
      //@ts-ignore
      dataSources.workshopsAPI.makeReservation(null, null)
    ).rejects.toThrow("No student_id");
  });

  it("fails on null option_id", async () => {
    await expect(
      //@ts-ignore
      dataSources.workshopsAPI.makeReservation("1", null)
    ).rejects.toThrow("No option_id");
  });
});

describe("api.getStudentReservation", () => {
  type ReservationForStudent = WorkshopReservation & {
    student: Student;
    option: WorkshopOption & { workshop: Workshop; teacher: Teacher };
  };
  test("gets a reservation", async () => {
    const student_id = "1";
    
    const mockValue: ReservationForStudent = {
      id: 1,
      create_time: new Date(2022,9,15),
      student_id: 1, 
      student: {
        id: 3,
        applicantId: 1,
        ciclo_actual: '2022B',
        codigo: "1234567890",
        nivel: 4,
        situacion: null,
        curso: "en",
        groupId: 2,
        desertor: false
      },
      option_id: 1,
      option: {
        ...option_1,
        teacher: teacher_1,
        workshop: workshop_1
      },
      attended: false
    };

    //@ts-ignore
    prismaMock.workshopReservation.findFirst.mockResolvedValue(mockValue);

    const reservation = await dataSources.workshopsAPI.getStudentReservation(1);

    expect(reservation?.student_id).toBe(1);
    expect(reservation).toMatchSnapshot({
      create_time: expect.any(Date),
    });
  });
});

describe("student", () => {
  const applicantObj = {
    id: 1,
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    carrera: "Abogado (DECH)",
    ciclo_ingreso: "2022B",
    telefono: "0001234567",
    email: "bad@bunny.pr",
    institutional_email: "benito.martinez@alumnos.udg.mx",
    externo: false,
  };

  const groupObject = {
    id: 2,
    ciclo: "2022B",
    name: "E4-1",
    time: "08:00-09:00",
    aula: "F4",
    nivel: 4,
    registrados: 30,
    course: "en",
    teacherId: 1,
  };

  const student_mock = {
    id: 3,
    codigo: "1234567890",
    nivel: 4,
    curso: "en",
    situacion: null,
    desertor: false,
    applicantId: 1,
    ciclo_actual: "2022B",
    groupId: 2,
  };

  it("gets a student with all applicant info", async () => {
    const codigo = "1234567890";
    const ciclo_actual = "2022B";

    const mockedValue: Awaited<ReturnType<StudentsAPI["getStudent"]>> = {
      ...student_mock,
      applicant: applicantObj,
      groupObject: groupObject,
    };
    //@ts-ignore
    prismaMock.student.findFirst.mockResolvedValue(mockedValue);

    const student = await dataSources.studentsAPI.getStudent(
      codigo,
      ciclo_actual
    );

    expect(student.codigo).toBe(codigo);
    expect(student.applicant.nombre).toBe("Benito Antonio");
    expect(student).toMatchSnapshot();
  });

  it("prisma returns a null applicant", async () => {
    //@ts-ignore
    prismaMock.student.findFirst.mockResolvedValue(student_mock);

    await expect(
      dataSources.studentsAPI.getStudent("1234567890", "2022B")
    ).rejects.toThrow("Applicant info is missing on database");
  });

  it("fails on null codigo", async () => {
    await expect(
      //@ts-ignore
      dataSources.studentsAPI.getStudent(null, null)
    ).rejects.toThrow("Codigo not provided on studentsAPI.getStudent");
  });

  it("fails on null ciclo_actual", async () => {
    await expect(
      //@ts-ignore
      dataSources.studentsAPI.getStudent("1234567890", null)
    ).rejects.toThrow("ciclo_actual not provided on studentsAPI.getStudent");
  });
});
