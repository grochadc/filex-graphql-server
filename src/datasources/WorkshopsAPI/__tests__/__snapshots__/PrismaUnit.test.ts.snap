// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api.getStudentReservation gets a reservation 1`] = `
{
  "attended": false,
  "create_time": Any<Date>,
  "id": "1",
  "option": {
    "active": true,
    "day": "Lunes",
    "id": 1,
    "place": "Centro de Aprendizaje Global",
    "teacher": {
      "id": 1,
      "nombre": "Gonzalo Rocha",
    },
    "teacher_id": 1,
    "time": "13:00",
    "url": null,
    "workshop": {
      "description": "Conversation Club",
      "id": 1,
      "levels": [
        1,
        2,
        3,
        4,
        5,
        6,
      ],
      "name": "Conversation",
    },
    "workshop_id": 1,
  },
  "option_id": 1,
  "student": {
    "applicantId": 1,
    "ciclo_actual": "2022B",
    "codigo": "1234567890",
    "curso": "en",
    "desertor": false,
    "groupId": 2,
    "id": 3,
    "nivel": 4,
    "situacion": null,
  },
  "student_id": 1,
}
`;

exports[`api.makeReservation makes a reservation 1`] = `
{
  "attended": false,
  "create_time": Any<Date>,
  "id": "1",
  "option": {
    "active": true,
    "day": "Lunes",
    "id": 1,
    "place": "Centro de Aprendizaje Global",
    "teacher": {
      "id": 1,
      "nombre": "Gonzalo Rocha",
    },
    "teacher_id": 1,
    "time": "13:00",
    "url": null,
    "workshop": {
      "description": "Conversation Club",
      "id": 1,
      "levels": [
        1,
        2,
        3,
        4,
        5,
        6,
      ],
      "name": "Conversation",
    },
    "workshop_id": 1,
  },
  "option_id": 1,
  "student": {
    "applicantId": 1,
    "ciclo_actual": "2022A",
    "codigo": "1234567890",
    "curso": "en",
    "desertor": false,
    "groupId": 1,
    "id": 1,
    "nivel": 3,
    "situacion": "AC",
  },
  "student_id": 1,
}
`;

exports[`student gets a student with all applicant info 1`] = `
{
  "applicant": {
    "apellido_materno": "Ocasio",
    "apellido_paterno": "Martinez",
    "carrera": "Abogado (DECH)",
    "ciclo_ingreso": "2022B",
    "codigo": "1234567890",
    "email": "bad@bunny.pr",
    "externo": false,
    "genero": "M",
    "id": 1,
    "institutional_email": "benito.martinez@alumnos.udg.mx",
    "nombre": "Benito Antonio",
    "telefono": "0001234567",
  },
  "applicantId": 1,
  "ciclo_actual": "2022B",
  "codigo": "1234567890",
  "curso": "en",
  "desertor": false,
  "groupId": 2,
  "groupObject": {
    "aula": "F4",
    "ciclo": "2022B",
    "course": "en",
    "id": 2,
    "name": "E4-1",
    "nivel": 4,
    "registrados": 30,
    "teacherId": 1,
    "time": "08:00-09:00",
  },
  "id": 3,
  "nivel": 4,
  "situacion": null,
}
`;
